src/lib/operators/sort.cpp

```diff
@@ -33,19 +20,77 @@ const std::string& Sort::name() const {
-      _output_chunk_size);
-  return _impl->_on_execute();
+  _sort_definition_data_types.resize(_sort_definitions.size());
+  std::transform(
+      _sort_definitions.begin(), _sort_definitions.end(), _sort_definition_data_types.begin(),
+      [this](const auto sort_definition) { return input_table_left()->column_data_type(sort_definition.column); });
+
+  std::shared_ptr<const Table> output = input_table_left();
+  std::shared_ptr<Table> output_mutable;
+  std::shared_ptr<PosList> previously_sorted_pos_list = std::shared_ptr<PosList>(nullptr);
+
+  _validate_sort_definitions();
+
+  for (auto column_id = _sort_definitions.size(); column_id-- != 0;) {
+    const bool is_last_sorting_run = (column_id == 0);
+
+    auto sort_definition = _sort_definitions[column_id];
+    DataType data_type = _sort_definition_data_types[column_id];
+
+    resolve_data_type(data_type, [&](auto type) {
+      using ColumnDataType = typename decltype(type)::type;
+
+      auto sort_single_col_impl = SortImpl<ColumnDataType>(input_table_left(), sort_definition.column,
+                                                           sort_definition.order_by_mode, _output_chunk_size);
+      const auto row_id_value_vector = sort_single_col_impl.sort_one_column(previously_sorted_pos_list);
+
+      if (is_last_sorting_run) {
+        auto materialization = std::make_shared<SortImplMaterializeOutput<ColumnDataType>>(
+            input_table_left(), row_id_value_vector, _output_chunk_size);
+
+        output_mutable = materialization->execute();
+      } else {
+        previously_sorted_pos_list.reset();
+        previously_sorted_pos_list = std::make_shared<PosList>();
+
+        for (auto& row_id_value : *row_id_value_vector) {
+          previously_sorted_pos_list->emplace_back(row_id_value.first);
+        }
+      }
+    });
+  }
+
+  auto final_sort_definition = _sort_definitions[0];
+  // Set the ordered_by attribute of the output's chunks to the column the chunk was sorted by last (because this
+  // is the column that the chunk is ordered by in any case; whereas it is not necessarily sorted by the other
+  // columns -- on their own -- as well).
+  const auto chunk_count = output_mutable->chunk_count();
+  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+    output_mutable->get_chunk(chunk_id)->set_ordered_by(
+        std::make_pair(final_sort_definition.column, final_sort_definition.order_by_mode));
+  }
+  output = output_mutable;
+
+  return output;
 }
 
-void Sort::_on_cleanup() { _impl.reset(); }
+void Sort::_on_cleanup() {}
+
+/**
+ * Asserts that all column definitions are valid.
+ */
+void Sort::_validate_sort_definitions() const {
+  const auto input_table = input_table_left();
+  for (const auto& column_sort_definition : _sort_definitions) {
+    Assert(column_sort_definition.column != INVALID_COLUMN_ID, "Sort: Invalid column in sort definition");
```

Comment:
Checking that the column id is within input_table->column_count() would be even safer.

Reply:
Done in 2e4a1e739bcc9286a783cba00336ca8e61325125